distributed/process.py

```diff
@@ -90,9 +104,42 @@ def _on_exit(self, exitcode):
             self._exit_callback(self)
         self._exit_future.set_result(exitcode)
 
+
+    @classmethod
+    def _immediate_exit_when_closed(cls, parent_alive_pipe):
+        """
+        Immediately exit the process when parent_alive_pipe is closed.
+        """
+        def monitor_parent():
+            try:
+                # The parent_alive_pipe should be held open as long as the
+                # parent is alive and wants us to stay alive. Nothing writes to
+                # it, so the read will block indefinitely.
+                parent_alive_pipe.recv()
+            except EOFError:
+                # Parent process went away unexpectedly. Exit immediately. Could
+                # consider other exiting approches here. My initial preference
+                # is to unconditionally and immediately exit. If we're in this
+                # state it is possible that a "clean" process exit won't work
+                # anyway - if, for example, the system is getting bogged down
+                # due to the running out of memory, exiting sooner rather than
+                # later might be needed to restore normal system function.
+                # If this is in appropriate for your use case, please file a
+                # bug.
+                os._exit(-1)
```

Comment:
Hm. One issue with this: the thing sending the signals will also receive it. So it will receive the `SIGINT`, and not likely run the `SIGKILL` later. I could go the route of masking `SIGINT` with `signal.pthread_kill`, but then we stray from portability. Thoughts?